src/librustc/middle/resolve_lifetime.rs

```diff
@@ -1673,6 +1673,16 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                         }
                         _ => None,
                     } {
+                        if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                            if !format!("{}", snippet).contains(&format!("{}", name)) {
+                                // Avoid spurious warnings: If the named lifetime
+                                // is not found in the snippit, then what would a person
+                                // getting this warning be able to do about it?
+                                debug!("skipping unused lifetime lint as not in snippit");
```

Comment:
I would really rather prefer we don't use stringification as a method of detecting whether to emit a lint, since it's both likely to be quite slow and is somewhat "not nice." Maybe we can instead check if the lint is being emitted within an async fn desugaring? That'll overapproximate and silence more than is necessary, but maybe that's better for now?

Reply:
That seems very overapproximate tho? `async fn foo<'a>() {}` would not complain about `'a` as AIUI.

Reply:
This should just check whether the second character of `ident` (the first after `'`) is "_". This is consistent with other `unused` lints and should fix this issue.

Reply:
Good point about performance - I'd not accounted that this was on a hot path.
I'll add some tests in that include macros also as I want to see how they interact with the warning.

At least we seem to be agreed that stringification, while slightly naff, is one potential solution. Let me see if there's a subtler approach...